### ISSUE TITLE
allow nested private classes to have public constructors

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#10161]
detected at: https://github.com/checkstyle/checkstyle/pull/9904#discussion_r655691910

it is not good that static analysis tools force us to make bad design of class. PMD is right. Checkstyle is not right.
If we agree that Chekcstyle is not right, we can register issue on this Check (if not yet already created).

in scope of this issue: make ctor public and lets suppress violation by Xpath
https://checkstyle.org/config_filters.html#SuppressionXpathSingleFilter , nested classes should allow public scope.

scope of problem:
```
✔ ~/java/github/romani/checkstyle [master] 
$ ag -Q "/* package */" src/main
src/main/java/com/puppycrawl/tools/checkstyle/PropertyCacheFile.java
396:        /* package */ ExternalResource(String location, String contentHashSum) {

src/main/java/com/puppycrawl/tools/checkstyle/filters/CsvFilterElement.java
49:    /* package */ CsvFilterElement(String pattern) {

src/main/java/com/puppycrawl/tools/checkstyle/filters/IntMatchFilterElement.java
35:    /* package */ IntMatchFilterElement(int matchValue) {

src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionCommentFilter.java
630:        /* package */ Tag(int line, int column, String text, TagType tagType,

src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithPlainTextCommentFilter.java
527:        /* package */ Suppression(

src/main/java/com/puppycrawl/tools/checkstyle/filters/IntRangeFilterElement.java
42:    /* package */ IntRangeFilterElement(int lowerBound, int upperBound) {

src/main/java/com/puppycrawl/tools/checkstyle/JavadocDetailNodeParser.java
747:        /* package */ ParseErrorMessage(int lineNumber, String messageKey,

src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java
442:        /* package */ FollowingEnumeration(NodeInfo start) {

src/main/java/com/puppycrawl/tools/checkstyle/ConfigurationLoader.java
490:        /* package */ InternalLoader()

src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilter.java
529:        /* package */ Tag(String text, int line, SuppressWithNearbyCommentFilter filter) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java
661:        /* package */ ResourceBundle(String baseName, String path, String extension) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/MultilineDetector.java
69:    /* package */ MultilineDetector(DetectorOptions options) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/SinglelineDetector.java
53:    /* package */ SinglelineDetector(DetectorOptions options) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/CommentSuppressor.java
39:    /* package */ CommentSuppressor(FileContents currentContents) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java
353:        /* package */ ClassContext(String className, DetailAST ast) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java
695:        /* package */ Values(BigInteger valueOfRange, BigInteger valueOfExpression) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java
354:        /* package */ Context(boolean checking) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/MethodCountCheck.java
401:        /* package */ MethodCounter(DetailAST scopeDefinition) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/ExecutableStatementCountCheck.java
253:        /* package */ Context(DetailAST ast) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java
381:        /* package */ ClassDesc(String qualifiedName, boolean declaredAsFinal,

src/main/java/com/puppycrawl/tools/checkstyle/checks/design/HideUtilityClassConstructorCheck.java
206:        /* package */ Details(DetailAST ast) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java
587:        /* package */ Entry(String checkName, int firstLine, int firstColumn,

src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsAvoidNullCheck.java
593:        /* package */ FieldFrame(FieldFrame parent) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java
1450:        /* package */ ClassFrame(AbstractFrame parent, DetailAST ident) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/AbstractSuperCheck.java
216:        /* package */ MethodNode(DetailAST ast) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java
342:        /* package */ Context(boolean checking) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java
888:        /* package */ FinalVariableCandidate(DetailAST variableIdent) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/HiddenFieldCheck.java
810:        /* package */ FieldFrame(FieldFrame parent, boolean staticType, String frameName) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/PkgImportControl.java
66:    /* package */ PkgImportControl(String packageName, boolean regex,
93:    /* package */ PkgImportControl(PkgImportControl parent, String subPackageName, boolean regex,

src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ClassImportRule.java
39:    /* package */  ClassImportRule(final boolean allow, final boolean localOnly,

src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/FileImportControl.java
44:    /* package */ FileImportControl(PkgImportControl parent, String name, boolean regex) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java
1307:        /* package */ ImportDetails(String importFullPath, String importGroup, boolean staticImport,
1398:        /* package */ RuleMatchForImport(String group, int length, int position) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/PkgImportRule.java
43:    /* package */ PkgImportRule(final boolean allow, final boolean localOnly,

src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java
1187:        /* package */ Token(String text, int lineNo, int columnNo) {
1198:        /* package */ Token(FullIdent fullIdent) {
1238:        /* package */ ExceptionInfo(DetailAST ast, ClassInfo classInfo) {

src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/HtmlTag.java
59:    /* package */ HtmlTag(String id, int lineNo, int position, boolean closedTag,

src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/TagParser.java
57:    /* package */ TagParser(String[] text, int lineNo) {
301:        /* package */ Point(int lineNo, int columnNo) {

src/main/java/com/puppycrawl/tools/checkstyle/gui/TreeTableCellRenderer.java
56:    /* package */ TreeTableCellRenderer(TreeTable treeTable, TreeModel model) {

src/main/java/com/puppycrawl/tools/checkstyle/gui/ListToTreeSelectionModelWrapper.java
46:    /* package */ ListToTreeSelectionModelWrapper(TreeTable jTreeTable) {

```

**Comment by romani** [comment]
@rnveach , @pbludov , @strkkk , please share your thought on this issue.

**Comment by pbludov** [comment]
From the compiler's point of view, any modifier of a private class member has the same access.

The difference will appear when the outer class is divided and there is a need for shared access to the inner class between several new classes. In such a scenario, for all members of the class that have explicit access modifier `private`, this access should be changed. 

On the other hand, having more public access to class members from the class itself can be confusing. For example, if the grep shows something like
```java
        public Inner() { ... }
```
This may be actually
```java
public class Outer {
  private class Inner {

       // ...

        public Inner() { ... }
   }
}
```

**Comment by romani** [comment]
@pbludov , I tried to read your post several times and I do not understand it. Please update it to be more precise.

My question was:
Ctor in nested classes should be allowed to have public scope. 
If team agree, we:
- create issue on Check to not violate such cases.
- temporary create xpath suppression for such violations, and update all cases to have public modifier.

**Comment by pbludov** [comment]
> Ctor in nested classes should be allowed to have public scope.

I want to say that both options have advantages and disadvantages. If we need to choose exactly one, I prefer the `public` scope.

**Comment by Stephan202** [comment]
IIUC, a variant of this issue is the following case:
```
$ javac PackagePrivateClass.java
$ cat -n PackagePrivateClass.java 
     1  final class PackagePrivateClass {
     2    public abstract class AbstractServiceLoadableClass {
     3      AbstractServiceLoadableClass(Object arg) {}
     4    }   
     5   
     6    public final class PublicClass extends AbstractServiceLoadableClass {
     7      public PublicClass() {
     8       super(new Object());
     9      }   
    10    }   
    11  }
$ cat -n config.xml
     1  <!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
     2  <module name="Checker">
     3      <module name="TreeWalker">
     4          <module name="RedundantModifier" />
     5      </module>
     6  </module>
$ java -jar checkstyle-10.3.3-all.jar -c config.xml PackagePrivateClass.java
Starting audit...
[ERROR] /tmp/repro/PackagePrivateClass.java:7:5: Redundant 'public' modifier. [RedundantModifier]
Audit done.
Checkstyle ends with 1 errors.
```

In my concrete case `PublicClass` is meant to be service-loadable, so the constructor must be public. @romani if you think that this is a separate issue then I don't mind opening a separate issue for it.

**Comment by romani** [comment]
>meant to be service-loadable

Please share what exact you mean by this. The more details the better, it is not only for me but for any other person who questioning such reasons.

**Comment by Stephan202** [comment]
@romani I'm referring to Java's [service loading](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html) mechanism, which [requires](https://github.com/openjdk/jdk/blob/739fdec7ea44be7ab12fdf418793276d581e58fe/src/java.base/share/classes/java/util/ServiceLoader.java#L654-L682) that types of instances loaded through [`ServiceLoader.load(Class)`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html#load(java.lang.Class)) are public and have a public constructor.

That said, I did misspeak in my previous message: while the class I'm having an issue with is of a type that is _generally_ service-loaded, it isn't in the exact context that triggered my post. My concrete use case is a test-scoped Error Prone `@BugChecker` very similar to [this one](https://github.com/google/error-prone/blob/1bbda1e7edd4028f2551b1a5194eaa0888a8e356/core/src/test/java/com/google/errorprone/bugpatterns/apidiff/ApiDiffCheckerTest.java#L79-L89), which requires a public constructor due to [this logic](https://github.com/google/error-prone/blob/1bbda1e7edd4028f2551b1a5194eaa0888a8e356/check_api/src/main/java/com/google/errorprone/scanner/ScannerSupplierImpl.java#L81-L95). (Checktyle's `RedundantModifier` check wouldn't flag that particular code, because the enclosing JUnit 4 test class is public, but in my case the `@BugChecker` is defined in a package-private JUnit 5 test class, which does cause the constructor to be flagged.)

I hope thing clears things up :). If not: the code I'm working on is scheduled to be open-sourced in ~three weeks, so once that happens I can just add a link here.

**Comment by romani** [comment]
we did some fixed to remove `package` visibility in our code in scope of - https://github.com/checkstyle/checkstyle/issues/12241.

But this issue is still valid and Check should be updated.

**Comment by Stephan202** [comment]
> If not: the code I'm working on is scheduled to be open-sourced in ~three weeks, so once that happens I can just add a link here.

It's been more than three weeks, but for completeness: [this](https://github.com/PicnicSupermarket/error-prone-support/blob/aa5ad4d25b91e8326729a2abd13fb52112c984ae/refaster-support/src/test/java/tech/picnic/errorprone/refaster/matchers/IsCharacterTest.java#L56-L61) is an example. Other cases can be found using [this](https://github.com/PicnicSupermarket/error-prone-support/search?q=%22checkstyle%2Fcheckstyle%2Fissues%2F10161%22) search query.

**Comment by DakshRJain737** [comment]
> we did some fixed to remove `package` visibility in our code in scope of - [#12241](https://github.com/checkstyle/checkstyle/issues/12241).
> 
> But this issue is still valid and Check should be updated.

Should all the constructor modifiers must be replaced with `public` along with the fix??
